File: 0085-maximal-rectangle/0085-maximal-rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int rows = matrix.size();
        if (rows == 0) return 0;
        int cols = matrix[0].size();

        // Count consecutive ones from the left
        vector<vector<int>> onesFromLeft(rows, vector<int>(cols, 0));
        for (int i = 0; i < rows; i++) {
            int run = 0;
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == '1') {
                    run++;
                    onesFromLeft[i][j] = run;
                } else {
                    run = 0;
                }
            }
        }

        vector<vector<int>> highBoundery(rows, vector<int>(cols, 0));
        vector<vector<int>> lowBoundery(rows, vector<int>(cols, 0));
        stack<int> st;

        // Upper boundary
        for (int col = 0; col < cols; col++) {
            while (!st.empty()) st.pop();

            for (int row = 0; row < rows; row++) {
                int cur = onesFromLeft[row][col];
                if (cur == 0) {
                    st.push(row);
                    continue;
                }

                while (!st.empty() && onesFromLeft[st.top()][col] >= cur)
                    st.pop();

                highBoundery[row][col] = st.empty() ? 0 : st.top() + 1;
                st.push(row);
            }
        }

        // Lower boundary
        for (int col = 0; col < cols; col++) {
            while (!st.empty()) st.pop();

            for (int row = rows - 1; row >= 0; row--) {
                int cur = onesFromLeft[row][col];
                if (cur == 0) {
                    st.push(row);
                    continue;
                }

                while (!st.empty() && onesFromLeft[st.top()][col] > cur)
                    st.pop();

                lowBoundery[row][col] = st.empty() ? rows : st.top();
                st.push(row);
            }
        }

        // Compute maximal rectangle
        int maxRectangle = 0;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int width = onesFromLeft[i][j];
                if (width == 0) continue;

                int height = lowBoundery[i][j] - highBoundery[i][j];
                maxRectangle = max(maxRectangle, width * height);
            }
        }

        return maxRectangle;
    }
};

File: 0085-maximal-rectangle/0085-maximal-rectangle.py
class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]:
            return 0

        m = len(matrix)
        n = len(matrix[0])

        # local bindings for speed
        width = [[0] * n for _ in range(m)]
        matrix0 = matrix[0]

        # compute consecutive widths
        for i in range(m):
            row = matrix[i]
            w = width[i]
            w[0] = 1 if row[0] == '1' else 0
            for j in range(1, n):
                if row[j] == '1':
                    w[j] = w[j - 1] + 1
                else:
                    w[j] = 0

        # compute up-span
        up = [[0] * n for _ in range(m)]
        for j in range(n):
            stack = []
            for i in range(m):
                cur_width = width[i][j]
                while stack and width[stack[-1]][j] > cur_width:
                    stack.pop()

                up[i][j] = i + 1 if not stack else i - stack[-1]
                stack.append(i)

        # compute down-span
        down = [[0] * n for _ in range(m)]
        for j in range(n):
            stack = []
            for i in range(m - 1, -1, -1):
                cur_width = width[i][j]
                while stack and width[stack[-1]][j] >= cur_width:
                    stack.pop()

                down[i][j] = (m - i) if not stack else (stack[-1] - i)
                stack.append(i)

        # compute maximal rectangle
        maxRect = 0
        for i in range(m):
            w_row = width[i]
            up_row = up[i]
            down_row = down[i]
            for j in range(n):
                h = up_row[j] + down_row[j] - 1
                area = w_row[j] * h
                if area > maxRect:
                    maxRect = area

        return maxRect
